core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java

```diff
@@ -48,11 +48,22 @@
     private Logger logger;
     private LogLevel requestLogLevel = LogLevel.DEBUG;
     private LogLevel successfulResponseLogLevel = LogLevel.DEBUG;
-    private LogLevel failedResponseLogLevel = LogLevel.WARN;
+    private LogLevel clientFailureResponseLogLevel = LogLevel.WARN;
+    private LogLevel serverFailureResponseLogLevel = LogLevel.WARN;
     private Function<? super RequestOnlyLog, LogLevel> requestLogLevelMapper =
             log -> requestLogLevel();
     private Function<? super RequestLog, LogLevel> responseLogLevelMapper =
-            log -> log.responseCause() == null ? successfulResponseLogLevel() : failedResponseLogLevel();
+            log -> {
+                if (log.responseCause() == null) {
+                    return successfulResponseLogLevel();
+                }
+
+                if (log.responseStatus().isClientError()) {
```

Comment:
How about using `Predicate`? 🤔 
It seems that it can be used in a variety of ways as well as server and client errors.

Reply:
Do you mean accepting a `Predicate` and `LogLevel`?

IMHO, a user can use `responseLogLevelMapper` if he/she wants to customize mapping more exquisitely. If don't, `clientFailureResponseLogLevel` and `serverFailureResponseLogLevel` would be useful shortcuts.

Reply:
gRPC or Thrift usually returns 200 status, although the request was failed and `log.responseCause()` is not null.
`serverFailureResponseLogLevel` will be used even if the HTTP status is not 5xx.
How about taking HTTP status and HTTP status class when setting `failureResponseLogLevel`.
```java
failureResponseLogLevel(HttpStatus, failedResponseLogLevel)
failureResponseLogLevel(HttpStatusClass, failedResponseLogLevel)

// Non matched status can follow `failureResponseLogLevel()` or the default.
LoggingService.builder()
              .failureResponseLogLevel(NOT_FOUND, DEBUG)
              .failureResponseLogLevel(UNAUTHORIZED, INFO)
              .failureResponseLogLevel(BAD_REQUEST, INFO)
              .failureResponseLogLevel(HttpStatusClass.SERVER_ERROR, ERROR);
              .failureResponseLogLevel(WARN);
 
```

Reply:
Sorry for the late response. Thanks.

```java
failureResponseLogLevel(HttpStatus, failedResponseLogLevel)
failureResponseLogLevel(HttpStatusClass, failedResponseLogLevel)
```

looks good to me.

```java
LoggingService.builder()
              .failureResponseLogLevel(NOT_FOUND, DEBUG)
              .failureResponseLogLevel(UNAUTHORIZED, INFO)
              .failureResponseLogLevel(BAD_REQUEST, INFO)
              .failureResponseLogLevel(HttpStatusClass.SERVER_ERROR, ERROR);
              .failureResponseLogLevel(WARN);
```

In this case, which behavior do you expect? Mapping LogLevel by matching in series of `failureResponseLogLevel` settings? E.g. `NOT_FOUND` -> `DEBUG`, `BAD_REQUEST` -> `INFO`, `INTERNAL_SERVER_ERROR` -> `ERROR`, `CONFLICT` (not matched) -> `WARN`?

Do I have to consider the following settings?

```java
LoggingService.builder()
              .failureResponseLogLevel(NOT_FOUND, DEBUG)
              .failureResponseLogLevel(WARN)
              .failureResponseLogLevel(UNAUTHORIZED, INFO); // which LogLevel to map?
```

Reply:
Like @heowc's idea, we can use `Predicate` by changing `LogLevelMapper`:

```java
interface LogLevelMapper extends Function<RequestLog, @Nullable LogLevel> {
  LogLevel andThen(LogLevelMapper another) {
    ...
  }
}

class HttpStatusLogLevelMapper { ... }
class HttpStatusClassLogLevelMapper { ... }

failureResponseLogLevel(failedResponseLogLevel) {
  return logLevelMapper(log -> {
    if (log.responseCause() != null) {
      return failedResponseLogLevel;
    }
    return null;
  });
}

failureResponseLogLevel(status, failedResponseLogLevel) {
  return logLevelMapper(new HttpStatusLogLevelMapper(status, failedResponseLogLevel));
}

failureResponseLogLevel(statusClass, failedResponseLogLevel) {
  return logLevelMapper(new HttpStatusClassLogLevelMapper(statusClass, failedResponseLogLevel));
}

logLevelMapper(logLevelMapper) {
  this.logLevelMapper = this.logLevelMapper.andThen(logLevelMapper);
}

```

But, I'm not sure this is a good idea since it's too complicated. This implementation may remove `isResponseLogLevelMapperSet` and `isResponseLogLevelSet` restrictions.

Reply:
> In this case, which behavior do you expect? Mapping LogLevel by matching in series of failureResponseLogLevel settings? 

Yeah, I think we can apply the mapper in the order of what they specified as we do in `RetryRule`.
So how about internally converting the `reponseLogLevel` to a mapper and applying them in the order?
```java
LoggingService.builder()
              .responseLogLevel(NOT_FOUND, DEBUG) // This will be set to responseLogLevelMapper.
                                                  // requestLog -> isNotFound(requestLog) ? return DEBUG : null;
              .responseLogLevel(OK, TRACE)
              .successfulResponseLogLevel(DEBUG) // successfulResponseLogLevel applied at the end.
              .failureResponseLogLevel(DEBUG)  // failureResponseLogLevel applied at the end.
              .responseLogLevelMapper(requestLog -> {
                  if (needsToWarn(requestLog)) {return WARN;}
                  return null; // return null to apply the next mapper.
              })
              .responseLogLevel(UNAUTHORIZED, INFO)
              .responseLogLevel(INTERNAL_SERVER_ERROR, INFO)
              .responseLogLevel(HttpStatusClass.SERVER_ERROR, WARN) // Everything will be warn except 500.
              
```

Reply:
Sorry, I'm late.
Otherwise, how about adding `LogLevelMapper` and `LogLevelMapperBuilder`?
```java
interface LogLevelMappler extends Function<? super RequestOnlyLog, LogLevel> {
    static LogLevelMapper of(HttpStatus, LogLevel) { ... }

    static LogLevelMaperBuilder builder() { ... }

    LogLevel apply(RequestLog requestLog);

    default LogLevelMapper orElse(LogLevelMapper other) {

    }
```

Reply:
I'd like the `LogLevelMapper`.
Could you tell me your intention about `Function<RequestOnlyLog, LogLevel>` and `LogLevel apply(RequestLog requestLog)`? Do you want to provide two apply methods that accept `RequestOnlyLog` and `RequestLog`?

Reply:
> I'm considering ResponseLogLevelMapper implementing Function<RequestLog, LogLevel> only.

You are right. It should be `Function<RequestLog, LogLevel>`. `ResponseLogLevelMapper` sounds good to me.